### ISSUE TITLE
Retry on some non-2xx HTTP status codes

### ISSUE THREAD

**Issue by rhodgkins** [sindresorhus/got#417]
Hi,

Just wanted to check why got doesn't do this...!

I thought the `retries` option would also be called HTTP errors but this doesn't appear to be the case ([Promise](https://github.com/sindresorhus/got/blob/v7.1.0/index.js#L182) / [as stream](https://github.com/sindresorhus/got/blob/v7.1.0/index.js#L250)) - is this by design?

I'd like got to retry certain 5xx error codes.

Cheers,

Rich

**Comment by petermetz** [comment]
https://github.com/sindresorhus/got/pull/421

**Comment by sindresorhus** [comment]
I think we should do this.

I propose is that we retry on the following methods and status codes by default:

Methods: `GET` `PUT` `HEAD` `DELETE` `OPTIONS` `TRACE`
Status codes: [`408`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408) [`413`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/413) [`429`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) [`500`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500) [`502`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/502) [`503`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503) [`504`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504)
*(Sourced from experience and other request clients)*

We also need to respect the [`Retry-After` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After).


Related issue: https://github.com/sindresorhus/got/issues/379

@lukechilds @reconbot @floatdrop @brandon93s Thoughts?

**Comment by reconbot** [comment]
Retry after support is awesome

**Comment by lukechilds** [comment]
One potential side effect to be aware of, especially if we're thinking of enabling this by default:

Consider you reach the request limit quota for the API you're using and they return `429 Too Many Requests` with a `Retry-After` header set for when your quota resets next month.

This would mean, if we by default retry `429`s and respect `Retry-After`, that making a request with Got to your API once you've hit your quota would return a Promise that may not resolve for up to 30 days in the future.

That doesn't seem very intuitive and could cause bugs that would be a nightmare to track down.

This seems like a great feature but might be better as opt in or not whitelisting `429`.

**Comment by reconbot** [comment]
I think automatically retrying is probably not a great idea in general,
this would be a good time to change that behavior.

**Comment by sindresorhus** [comment]
Networks are inherently unreliable. I think retrying by default makes sense, but we should be very conservative with the default.

**Comment by brandon93s** [comment]
> Methods: GET PUT HEAD DELETE OPTIONS TRACE

This list is sensible, as these methods _should_ be safe or idempotent.

Would this functionality be rolled up into the existing `retries` option, or separated for more control? For API simplicity it makes sense to combine them, but the proposed functionality is more likely to need disabling than the current implementation. 

> Status codes: 408 413 429 500 502 503 504

`413` may be an odd one to retry since the retry would be sent with an identical payload.

**Comment by sindresorhus** [comment]
> Would this functionality be rolled up into the existing retries option, or separated for more control? For API simplicity it makes sense to combine them

I was thinking of combining them, but also allow passing an object where people can customize:

```js
retries: {
	retry: Number | Function,
	methods: string[],
	statusCodes: number[]
}
```

Thoughts?

> but the proposed functionality is more likely to need disabling than the current implementation.

Why do you think so?

> 413 may be an odd one to retry since the retry would be sent with an identical payload.

I agree. Let's take it out.

**Comment by sindresorhus** [comment]
Actually, regarding `413`:

> 6.5.11.  413 Payload Too Large

>   The 413 (Payload Too Large) status code indicates that the server is
   refusing to process a request because the request payload is larger
   than the server is willing or able to process.  The server MAY close
   the connection to prevent the client from continuing the request.

>  ** If the condition is temporary, the server SHOULD generate a
   Retry-After header field to indicate that it is temporary and after
   what time the client MAY try again.**

> \- https://tools.ietf.org/html/rfc7231#section-6.5.11

So it seems we could handle `413` as long as it has a `Retry-After` header and the time specified there is below the set timeout. Am I interpreting it correctly?

**Comment by brandon93s** [comment]
Agreed, retry `413` if and only if (iff) the response provides a `Retry-After` header. `Retry-After` could probably serve as a base case to enable retry, regardless of error, for the HTTP methods mentioned before. 

> I was thinking of combining them, but also allow passing an object where people can customize

Looks good to me. This can also be made backwards compatible by mapping `retries Number | Function` to `retry.retries` when we normalize arguments. 

> Why do you think so?

The aforementioned errors _should_ be safe, but I've encountered some fairly non-standard APIs where I wouldn't want to retry on a `500` for example because I'm not confident the method is safe / idempotent. But, being able to specify `methods` and `statusCodes` solves for this.

**Comment by alixaxel** [comment]
@sindresorhus I'd like to use auto-retries on a 202 response (since it's async).

However, it seems that `statusCodes` only plays an effect if it's >= 400, is that right?

**Comment by robbieaverill** [comment]
Hi everyone, I came across this on a hunch that requests were automatically retried. @lukechilds said in [his comment](https://github.com/sindresorhus/got/issues/417#issuecomment-364760887) that:

>That doesn't seem very intuitive and could cause bugs that would be a nightmare to track down.

I have to say I entirely agree. It depends on how you're using this library, but I'd much rather tell my users that the request was rate limited than allow it to infinitely keep retrying. I've disabled it in my config, but thought I'd add my two cents here.

Thanks for creating this library, it's great!
